R/renamed-files.R

```diff
@@ -0,0 +1,151 @@
+#' Set the column  `changed_files` of a `log` to the latest name of the file
+#'
+#' [parse_log_detailed()] parses a git log. However, changed file names
+#' are not corrected retroactively. Hence, if one wants to process file-specific
+#' commit data for one file across different names for that file, the column
+#' `changed_files` has to be unified. This is done with this function by
+#' updating the file names recursively and setting them to the latest name used.
+#' @param log An unnested detailed log.
+#' @examples
+#' library("magrittr")
+#' gitsumlog %>%
+#'   tidyr::unnest() %>%
+#'   set_changed_file_to_latest_name() %>%
+#'   add_line_history()
+#' @export
+set_changed_file_to_latest_name <- function(log) {
+  is_renaming <- is_name_change(log$changed_file)
+  contains_renaming <- log[is_renaming, ]
+  contains_no_renaming <- log[!is_renaming,]
+  transition_model <- parse_reassignment(
+    contains_renaming$changed_file, which(is_renaming)
+  )
+  update_changed_file_sequentially(log, transition_model)
+}
+
+
+#' Detect whether a `changed_file`-entry is a renaming
+#'
+#' @param changed_file The column `changed_file` from a parsed unnested
+#'   detailed log.
+#' @param reassignment_pattern The regular expression pattern that corresponds
+#'   to renaming.
+#' @importFrom stringr str_locate
+#' @examples
+#' gitsum:::is_name_change("R/{gitsum.R => gitsum-package.R}")
+is_name_change <- function(changed_file,
+                           reassignment_pattern = "\\{.+\\s\\=\\>.+\\}$") {
+  regex_position_matrix <- str_locate(changed_file, reassignment_pattern)
+  apply(regex_position_matrix, 1, function(x) !all(is.na(x)))
+}
+
+#' Parse a raw reassignment into a list
+#'
+#' Takes a vector of raw reassignments and turns them into a list of elements.
+#' Every element contains the old path, the new path and the index of the
+#' reassignment.
+#' @param raw_reassignment Character vector with raw re-assignments.
+#' @param reassignment_index Integer vector indicating the position of the
+#'   re-assignment in the original unnested detailed log.
+#' @examples
+#' gitsum:::parse_reassignment(rep("R/{a => b}", 2), c(1, 2))
```

Comment:
What about renaming which results in a file moving between different sub-directories?

Reply:
I think the prefix before `{` will always be the "least common root", but there won't be braces if the file is in the root directory (https://github.com/lorenzwalthert/gitsum/pull/15#discussion_r160056711)